store/index/index.go

```diff
@@ -90,119 +92,226 @@ const BucketPoolSize = 1024
 //
 // It is created if there is no existing index at that path.
 func OpenIndex(path string, primary primary.PrimaryStorage, indexSizeBits uint8) (*Index, error) {
+	var lastIndexNum uint32
+	var lastIndexPath string
 	var file *os.File
-	var buckets Buckets
-	var sizeBuckets SizeBuckets
-	var length types.Position
-	stat, err := os.Stat(path)
-	if os.IsNotExist(err) {
-		header := FromHeader(NewHeader(indexSizeBits))
-		headerSize := make([]byte, 4)
-		binary.LittleEndian.PutUint32(headerSize, uint32(len(header)))
+	headerPath := path + ".info"
 
-		file, err = openFileRandom(path, os.O_RDWR|os.O_APPEND|os.O_EXCL|os.O_CREATE)
-		if err != nil {
-			return nil, err
-		}
-		if _, err := file.Write(headerSize); err != nil {
-			return nil, err
-		}
-		if _, err = file.Write(header); err != nil {
-			return nil, err
-		}
-		if err := file.Sync(); err != nil {
-			return nil, err
-		}
-		length = types.Position(len(header) + len(headerSize))
-		buckets, err = NewBuckets(indexSizeBits)
-		if err != nil {
+	buckets, err := NewBuckets(indexSizeBits)
+	if err != nil {
+		return nil, err
+	}
+	sizeBuckets, err := NewSizeBuckets(indexSizeBits)
+	if err != nil {
+		return nil, err
+	}
+	fileBuckets, err := NewFileBuckets(indexSizeBits)
+	if err != nil {
+		return nil, err
+	}
+
+	err = upgradeIndex(path, headerPath)
+	if err != nil {
+		return nil, err
+	}
+
+	header, err := readHeader(headerPath)
+	if os.IsNotExist(err) {
+		header = newHeader(indexSizeBits)
+		if err = writeHeader(headerPath, header); err != nil {
 			return nil, err
 		}
-		sizeBuckets, err = NewSizeBuckets(indexSizeBits)
+		lastIndexPath = indexFileName(path, 0)
+		file, err = openNewFileAppend(lastIndexPath)
 		if err != nil {
 			return nil, err
 		}
 	} else {
 		if err != nil {
 			return nil, err
 		}
-		buckets, sizeBuckets, err = scanIndex(path, indexSizeBits)
+
+		if header.BucketsBits != indexSizeBits {
+			return nil, types.ErrIndexWrongBitSize{header.BucketsBits, indexSizeBits}
+		}
+
+		lastIndexNum, err = scanIndex(path, header.FirstFile, indexSizeBits, buckets, sizeBuckets, fileBuckets)
 		if err != nil {
 			return nil, err
 		}
-		file, err = openFileRandom(path, os.O_RDWR|os.O_APPEND|os.O_EXCL)
+
+		lastIndexPath = indexFileName(path, lastIndexNum)
+		file, err = openFileAppend(lastIndexPath)
 		if err != nil {
 			return nil, err
 		}
```

Comment:
Refactor this `if` and `else` to instantiate a `var lastIndextPath` outside it? That way the following block, duplicated in branches, can also move out which maybe makes the code easier to read:

```go
file, err = openNewFileAppend(lastIndexPath)
if err != nil {
    return nil, err
}
```

Reply:
Done.  Thank you.